File: 2020/19/rules.py
import itertools


class Rule:
    def __init__(self, rule_string=None, value=None):
        if value:
            self.value = value
        elif rule_string[0] == '"':
            self.value = rule_string[1]
        else:
            self.value = [val.split() for val in rule_string.split(" | ")]

    def resolve(self, rules):
        if isinstance(self.value, str):
            return
        self.value = [[rules[val] for val in entry] for entry in self.value]

    def possible_lengths(self):
        if not hasattr(self, "_possible_lengths"):
            if isinstance(self.value, str):
                return {1}
            output = set()
            for rules in self.value:
                output = output.union(
                    set(
                        sum(prod)
                        for prod in itertools.product(
                            *[rule.possible_lengths() for rule in rules]
                        )
                    )
                )
            self._possible_lengths = output
        return self._possible_lengths

    def validate(self, message):
        if len(message) not in self.possible_lengths():
            return False
        if isinstance(self.value, str):
            return self.value == message
        valid = False
        for rules in self.value:
            for length in rules[0].possible_lengths():
                valid = rules[0].validate(message[:length])
                if valid and len(rules) > 1:
                    valid = valid and Rule(value=[rules[1:]]).validate(message[length:])
                elif len(message[length:]) > 0:
                    valid = False
                if valid:
                    return True
        return valid


with open("input.txt", "r") as fid:
    rules, messages = fid.read().split("\n\n")
    rules = {val.split(": ")[0]: Rule(val.split(": ")[1]) for val in rules.split("\n")}
    for rule in rules.values():
        rule.resolve(rules)
    ct = 0
    for message in messages.split("\n"):
        if rules["0"].validate(message):
            ct += 1
    print(ct)

File: 2020/19/rules.ts
import * as fs from 'fs';

function getInput(filename: string): [string[], string[]] {
  const input: string[] = fs.readFileSync(filename, 'utf8').split('\n\n');
  return [input[0].split('\n'), input[1].split('\n')];
}

class Rule {
  name: string

  value: string | string[][] | Rule[][]

  constructor(name: string, value: string | string[][] | Rule[][]) {
    this.name = name;
    this.value = value;
  }

  convert(lookup: {[ruleNumber: string]: Rule}): void {
    if (typeof this.value === 'string') {
      return;
    }
    let key: string;
    for (let i: number = 0; i < this.value.length; i += 1) {
      for (let j: number = 0; j < this.value[i].length; j += 1) {
        if (typeof this.value[i][j] === 'string') {
          key = this.value[i][j] as string;
          lookup[key].convert(lookup);
          this.value[i][j] = lookup[key];
        }
      }
    }
  }

  validate(input: string): boolean {
    if (typeof this.value === 'string') {
      return this.value === input;
    }
    let currentRules: Rule[];
    let valid: boolean = false;
    let tempRule: Rule;
    for (let i: number = 0; i < this.value.length; i += 1) {
      if (this.value[i].length > 0 && typeof this.value[i][0] === 'string') {
        throw new Error('convert before validating');
      }
      currentRules = this.value[i] as Rule[];
      for (let j: number = 1; j <= input.length - currentRules.length + 1; j += 1) {
        valid = currentRules[0].validate(input.slice(0, j));
        if (valid && currentRules.length > 1) {
          tempRule = new Rule('temp', [currentRules.slice(1)]);
          valid = valid && tempRule.validate(input.slice(j));
        } else if (input.slice(j).length > 0) {
          valid = false;
        }
        if (valid) {
          return true;
        }
      }
    }
    return valid;
  }
}

function getRules(input: string[]): {[ruleNumber: string]: Rule} {
  const rules: {[ruleNumber: string]: Rule} = {};
  let key: string;
  let remainder: string;
  input.forEach((rule: string): void => {
    [key, remainder] = rule.split(': ');
    if (remainder[0] === '"') {
      rules[key] = new Rule(key, remainder[1]);
    } else {
      rules[key] = new Rule(
        key,
        remainder.split(' | ').map((ruleSet: string): string[] => ruleSet.split(' ')),
      );
    }
  });
  return rules;
}

function countValidMessages(rule: Rule, messages: string[]): number {
  let count: number = 0;
  messages.forEach((message: string, i: number): void => {
    console.log(i);
    if (message.length === 24 && rule.validate(message)) {
      count += 1;
    }
  });
  return count;
}

const [rulesInput, messagesInput]: [string[], string[]] = getInput('input.txt');
const rules: {[ruleNumber: string]: Rule} = getRules(rulesInput);
rules[0].convert(rules);
console.log(countValidMessages(rules[0], messagesInput));
